Add media learning area. (#3215)
GoogleChrome/web.dev: src/site/content/en/media/video-and-source-tags/index.md

---
layout: post
title: The <video> and <source> tags
authors:
  - samdutton
  - joemedley
description: |
  You've properly prepared a video file for the web. You've given it correct

>   dimensions and the correct resolution. You've encrypted it. You've even

Reviewer: In the file prep section, we talk about encrypting video content, but we don't talk about playback of encrypted content anywhere.

  created separate webm and mp4 files for different browsers. For anyone to see
  it, you still need to add it to a web page.
date: 2014-14-15
updated: 2020-08-20

> ---

Reviewer 2: @jpmedley I plan on redirecting the old H5R guide (https://www.html5rocks.com/en/tutorials/video/basics/) to this page (`/video-and-source-tags`) once this PR lands. Can you check the old H5R guide and see if there are any useful/interesting ideas that should be incorporated into this new web.dev page?

Author: I'm incorporating this one too. Take it off your list. https://www.html5rocks.com/en/tutorials/track/basics/

You've properly [prepared a video file](prepare-media/) for the web. You've
given it correct dimensions and the correct resolution. You've encrypted it.
You've even created separate webm and mp4 files for different browsers.

For anyone to see it, you still need to add it to a web page. Doing so properly
requies two HTML elements: the `<video>` element and the `<source>`.  Because it
makes no sense to cover them separately, the page explains both. It also
explains attributes you should add to those tags to craft a good user
experience.

{% Aside %}
You always have the option of uploading your file to YouTube or Vimeo. In many
cases, this is preferable to the procedure described here. Those services
handle formatting and filetype conversion for you, as well as provide the means
to embedd a video in your web page. If need to manage this yourself, read on.
{% endAside %}

## Specify a single file

Although it's not recommended, you can use the video element by itself. Always

> use the `type` attribute as shown below. the browser uses this to determine if

Reviewer: ```suggestion
use the `type` attribute as shown below. The browser uses this to determine if
```

otherwise video source type checks won't work.
{% endAside %}

### Specify start and end times

Save bandwidth and make your site feel more responsive: use media fragments to
add start and end times to the video element.

<video controls>
  <source src="https://storage.googleapis.com/web-dev-assets/video-and-source-tags/chrome.webm#t=5,10" type="video/webm">
  <source src="https://storage.googleapis.com/web-dev-assets/video-and-source-tags/chrome.mp4#t=5,10" type="video/mp4">
  <p>This browser does not support the video element.</p>
</video>

To use a media fragment, add `#t=[start_time][,end_time]` to the media URL. For
example, to play the video from seconds 5 to 10, specify:

```html
<source src="video/chrome.webm#t=5,10" type="video/webm">
```

You can use this feature to deliver multiple views on the same video&ndash;like
cue points in a DVD&ndash;without having to encode and serve multiple files.

{% Aside 'caution' %}
By default, most servers enable range requests, which is required for media
fragments to work. Because some hosting services turn them off, you should test
this feature with your server before using it on a live site.
{% endAside %}

> Using your browser developer tools, check for `Accept-Ranges: bytes` in the

Reviewer: I found the context for this confusing. Is this to determine whether ranges are supported?

Author: This is related to the previous caution. I started to rework these two sections and the image into a range requests section. The problem is, it's not used anywhere else and as far as I know it's not relevant to videos in &lt;video> tags. What do you think about removing it?

Reviewer: Remove the range section completely? I think it's useful to have. It just wasn't clear to me that the paragraph was related to the caution.

{% Aside 'caution' %}
Don't force element sizing that results in an aspect ratio different from the
original video. Squashed or stretched looks bad.
{% endAside %}

**CSS:**

```css
.video-container {
    position: relative;
    padding-bottom: 56.25%;
    padding-top: 0;
    height: 0;
    overflow: hidden;
}

.video-container iframe,
.video-container object,
.video-container embed {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
}
```

**HTML:**

```html
<div class="video-container">
  <iframe src="//www.youtube.com/embed/l-BA9Ee2XuM"
          frameborder="0" width="560" height="315">
  </iframe>
</div>
```

[Try it](https://googlesamples.github.io/web-fundamentals/fundamentals/media/responsive_embed.html)

Compare the [responsive sample](https://googlesamples.github.io/web-fundamentals/fundamentals/media/responsive_embed.html)
to the [unresponsive version](https://googlesamples.github.io/web-fundamentals/fundamentals/design-and-ux/responsive/unyt.html).

### Device orientation

Device orientation isn't an issue for desktop monitors or laptops, but it's
hugely important when considering web page design for mobile devices and
tablets.

> Safari on iPhone do a good job of switching between portrait and landscape

Reviewer: ```suggestion
Safari on iPhone does a good job of switching between portrait and landscape
```

orientation:

<div class="w-columns">
<figure class="w-figure">
  <img src="./iphone-video-playing-portrait.png" alt="Screenshot of video playing in Safari on iPhone, portrait.">
  <figcaption class="w-figcaption">Screenshot of video playing in Safari on iPhone, portrait.</figcaption>
</figure>
<figure class="w-figure">
  <img src="./iphone-video-playing-landscape.png" alt="Screenshot of video playing in Safari on iPhone, landscape.">
  <figcaption class="w-figcaption">Screenshot of video playing in Safari on iPhone, landscape.</figcaption>
</figure>
</div>

> Device orientation on an iPad and Chrome on Android can be problematic.

Reviewer: Is this still true?

Author: I believe so. I just looked at a video on iPad safari that does not have `width: 100%` or `max-width: 100%`. It looked roughly like the image below. I could be missing something. I'll investigate further.

For example, without any customization a video playing on an iPad in landscape
orientation looks like this:

<figure class="w-figure">
  <img src="./ipad-retina-landscape-video-playing.png" alt="Screenshot of video playing in Safari on iPad Retina, landscape.">

>   <figcaption class="w-figcaption">Screenshot of video playing in Safari on iPad Retina, landscape.</figcaption>

Reviewer: ```suggestion
  <figcaption class="w-figcaption">Screenshot of video playing in Safari on iPad, landscape.</figcaption>
```

</figure>

Setting the video `width: 100%` or `max-width: 100%` with CSS can resolve
many device orientation layout problems.

> ### Autoplay

Reviewer: Is it worth mentioning restrictions on autoplay in Chrome? https://developers.google.com/web/updates/2017/09/autoplay-policy-changes

Reviewer: I think the policy is pretty stable now. Irrespective of whether it's controversial, it would be good to link to the article.

Author: At some point, we may wan to give this its own page. I thought that would hold up getting this out, so I came up with something that covers it with appropriate links.

Author: Link added below.

  video source without having to download part of the video to "sniff" the format.

These issues are especially important in mobile contexts, where bandwidth and
latency are at a premium and the user's patience is likely limited. Omitting the
type attribute can affect performance when there are multiple sources with
unsupported types.

{% Aside %} There are a few ways you can dig into the details. Check out [A
Digital Media Primer for Geeks](//www.xiph.org/video/vid1.shtml) to find out
more about how video and audio work on the web. You can also use [remote
debugging](https://developers.google.com/web/tools/chrome-devtools/remote-debugging)
in DevTools to compare network activity [with type
attributes](https://googlesamples.github.io/web-fundamentals/fundamentals/media/video-main.html)
and [without type
attributes](https://googlesamples.github.io/web-fundamentals/fundamentals/design-and-ux/responsive/notype.html).

Also check the response headers in your browser developer tools to
[ensure your server reports the right MIME type](//developer.mozilla.org/en/docs/Properly_Configuring_Server_MIME_Types);
otherwise video source type checks won't work.
{% endAside %}

> ### Specify start and end times

Author: @rkochman You had an earlier comment about range request material not making much sense. (Ignore my suggestion that it be removed.) I found some additional information about fragments, for which range requests are a prerequisite. I've improved the section  and improved how the concepts are tied to each other. Please read this whole section again.

Reviewer: I do like that--thanks!

---
layout: post
title: The <video> and <source> tags
authors:
  - samdutton
  - joemedley
description: |
  You've properly prepared a video file for the web. You've given it correct
  dimensions and the correct resolution. You've even created separate WebM and
  MP4 files for different browsers. For anyone to see it, you still need to add
  it to a web page.
date: 2014-14-15

> updated: 2020-08-20

Reviewer 2: ```suggestion
updated: 2020-08-20
tags:
  - media
  - video
```

---

You've properly [prepared a video file](../prepare-media/) for the web. You've
given it correct dimensions and the correct resolution. You've even created
separate WebM and MP4 files for different browsers.

For anyone to see it, you still need to add it to a web page. Doing so properly
requires two HTML elements: the
[`<video>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video)
element and the
[`<source>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/source).
In addition to basics about these tags, this section also explains attributes
you should add to those tags to craft a good user experience.

{% Aside %}
You always have the option of uploading your file to
[YouTube](https://www.youtube.com/) or [Vimeo](https://vimeo.com/). In many
cases, this is preferable to the procedure described here. Those services handle
formatting and filetype conversion for you, as well as provide the means to

> embedd a video in your web page. If need to manage this yourself, read on.

Reviewer 2: ```suggestion
embedd a video in your web page. If you need to manage this yourself, read on.
```

{% endAside %}

## Specify a single file

Although it's not recommended, you can use the video element by itself. Always
use the `type` attribute as shown below. The browser uses this to determine if
it can play the provided video file. If it can't, the enclosed text displays.

```html
<video src="chrome.webm" type="video/webm">
    <p>Your browser cannot play the provided video file.</p>
</video>
```

### Specify multiple file formats

Recall from [File basics](../file-basics) that not all browsers support the same
video formats. The `<source>` element lets you specify multiple formats as a
fallback in case the user's browser doesn't support one of them.

> The example below produces the ebedded video that immediately follows.

Reviewer 2: ```suggestion
The example below produces the embedded video that immediately follows.
```

```html
<video controls>
  <source src="https://storage.googleapis.com/web-dev-assets/video-and-source-tags/chrome.webm" type="video/webm">
  <source src="https://storage.googleapis.com/web-dev-assets/video-and-source-tags/chrome.mp4" type="video/mp4">

>     <p>Your browser cannot play the provided video file.</p>

Reviewer 2: ```suggestion
  <p>Your browser cannot play the provided video file.</p>
```
Fixing the indentation since it's a sibling element of the `source` elements

</video>
```

[Try it](https://track-demonstration.glitch.me) ([source](https://glitch.com/edit/#!/track-demonstration))

When the browser parses the `<source>` tags, it uses the optional `type`
attribute to determine which file to download and play. If the browser
supports WebM, it plays that; if not, it checks whether it can play

> MPEG-4 videos.

Reviewer 2: This is based on the order of the elements as specified in the HTML, right? I guess that may be obvious but I feel like it would be nice to explicitly mention.

Author: The browser determines the correct file based on the given (MIME) `type`. It never downloads what it can't play. That makes the order irrelevant when the `type` attribute is present. I haven't been able to learn what happens when there's no `type` attribute. The spec is silent on this point, which makes it non-normative (variable by browser) by default. I see no reason (in the spec) that a browser couldn't kick off simultaneous downloads.

This approach has several advantages over serving different HTML or server-side
scripting, especially on mobile:

* You can list formats in order of preference.
* Native client-side switching reduces latency; only one request is made to
  get content.
* Letting the browser choose a format is simpler, quicker, and potentially
  more reliable than using a server-side support database with user-agent detection.
* Specifying each file source's type improves network performance; the browser can select a
  video source without having to download part of the video to "sniff" the format.

These issues are especially important in mobile contexts, where bandwidth and
latency are at a premium and the user's patience is likely limited. Omitting the

> type attribute can affect performance when there are multiple sources with

Reviewer 2: ```suggestion
`type` attribute can affect performance when there are multiple sources with
```

unsupported types.

{% Aside %} There are a few ways you can dig into the details. Check out [A
Digital Media Primer for Geeks](//www.xiph.org/video/vid1.shtml) to find out
more about how video and audio work on the web. You can also use [remote
debugging](https://developers.google.com/web/tools/chrome-devtools/remote-debugging)
in DevTools to compare network activity [with type
attributes](https://googlesamples.github.io/web-fundamentals/fundamentals/media/video-main.html)
and [without type
attributes](https://googlesamples.github.io/web-fundamentals/fundamentals/design-and-ux/responsive/notype.html).

Also check the response headers in your browser developer tools to
[ensure your server reports the right MIME type](//developer.mozilla.org/en/docs/Properly_Configuring_Server_MIME_Types);
otherwise video source type checks won't work.
{% endAside %}

### Specify start and end times

Save bandwidth and make your site feel more responsive: use media fragments to
add start and end times to the video element.

> <video controls>
>   <source src="https://storage.googleapis.com/web-dev-assets/video-and-source-tags/chrome.webm#t=5,10" type="video/webm">
>   <source src="https://storage.googleapis.com/web-dev-assets/video-and-source-tags/chrome.mp4#t=5,10" type="video/mp4">
>   <p>This browser does not support the video element.</p>
> </video>

Reviewer 2: ```suggestion
<figure class="w-figure">
  <video controls class="w-screenshot">
    <source src="https://storage.googleapis.com/web-dev-assets/video-and-source-tags/chrome.webm#t=5,10" type="video/webm">
    <source src="https://storage.googleapis.com/web-dev-assets/video-and-source-tags/chrome.mp4#t=5,10" type="video/mp4">
    <p>This browser does not support the video element.</p>
  </video>
</figure>
```
This should make the video's styling a bit more polished

To use a media fragment, add `#t=[start_time][,end_time]` to the media URL. For
example, to play the video from seconds 5 to 10, specify:

```html
<source src="video/chrome.webm#t=5,10" type="video/webm">
```

> You can also specify the times in hours:minutes:seconds. For example,

Reviewer 2: ```suggestion
You can also specify the times in `<hours>:<minutes>:<seconds>`. For example,
```

> #t=00:01:05 starts the video at one minute, five seconds. To play only the first

Reviewer 2: ```suggestion
`#t=00:01:05` starts the video at one minute, five seconds. To play only the first
```

> minute of video, specify #t=,00:01:00

Reviewer 2: ```suggestion
minute of video, specify `#t=,00:01:00`.
```

You can use this feature to deliver multiple views on the same video&ndash;like
cue points in a DVD&ndash;without having to encode and serve multiple files.

For this feature to work, your server must support range requests and that
capability must be enabled. Most servers enable range requests by default.
Because some hosting services turn them off, you should confirm that range
requests are available for using fragments on your site. Fortunately, you can do
this in your brower developer tools.

> After uploading your media and your HTML page to your server, visit the page in
> a browser. Next, open DevTools by right-clicking and selecting **Inspect**.
> Select **Network** in the top menu. From the browser page itself, click reload.
> Back in DevTools, locate your media file in the **Name** column. When you click
> the file, the pages request and response headers should appear. I've shown this
> in the image below. Look for the `Accept-Ranges` header is and verify that is
> says `bytes`. In a the image, I've drawn a red box  around this header. If you
> do not see `butes` as the value, you'll need to contact your hosting provider.

Reviewer 2: ```suggestion
```
I think we can get rid of this and assume that the audience knows basic Chrome DevTools usage. We should keep the screenshot, though.

<figure class="w-figure">

>   <img src="./accept-ranges-chrome-devtools.png" alt="Chrome DevTools screenshot: Accept-Ranges: bytes.">

Reviewer 2: ```suggestion
  <img class="w-screenshot w-screenshot--filled" src="./accept-ranges-chrome-devtools.png" alt="Chrome DevTools screenshot: Accept-Ranges: bytes.">
```

>   <figcaption class="w-figcaption">Chrome DevTools screenshot: Accept-Ranges: bytes.</figcaption>

Reviewer 2: Please update this to provide a little more context (which is now needed assuming that you have agreed to delete the last paragraph).

Author: Done.

</figure>

### Include a poster image

Add a poster attribute to the `video` element so that viewers have an idea of
the content as soon as the element loads, without needing to download the video
or start playback.

```html
<video poster="poster.jpg" ...>

>   ...

Reviewer 2: ```suggestion
  …
```

  <img src="./chrome-android-video-poster.png" alt="A fallback poster makes it seem as if the first frame has been captured.">
</figure>

{% CompareCaption %}
A fallback poster makes it seem as if the first frame has been captured.
{% endCompareCaption %}

{% endCompare %}
</div>

### Ensure videos don't overflow containers

When video elements are too big for the viewport, they may overflow their
container, making it impossible for the user to see the content or use the
controls.

<div class="w-columns">
  <figure class="w-figure">
    <img src="./chrome-android-portrait-video-unstyled.png" alt="Android Chrome screenshot, portrait: unstyled video element overflows
    viewport.">
    <figcaption class="w-figcaption">Android Chrome screenshot, portrait: unstyled video element overflows
    viewport.</figcaption>
  </figure>
  <figure class="w-figure">
    <img src="./chrome-android-landscape-video-unstyled.png" alt="Android Chrome screenshot, landscape: unstyled video element overflows
    viewport.">
    <figcaption class="w-figcaption">Android Chrome screenshot, landscape: unstyled video element overflows
    viewport.</figcaption>
  </figure>
</div>

> You can control video dimensions using JavaScript or CSS. JavaScript libraries
> and plugins such as [FitVids](http://fitvidsjs.com/) (outside the scope fo this
> article) make it possible to maintain appropriate size and aspect ratio, even
> for videos from YouTube and other sources.
> 
> For simple uses like the ones I'm describing here, use [CSS media
> queries](/web/fundamentals/design-and-ux/responsive/#css-media-queries) to
> specify the size of elements depending on the viewport dimensions; `max-width:
> 100%` is your friend.

Reviewer 2: ```suggestion
You can control video dimensions using CSS. For simple uses like the ones
I'm describing here, use [CSS media
queries](/responsive-web-design-basics/#media-queries) to
specify the size of elements depending on the viewport dimensions; `max-width:
100%` is your friend.
```
Given that we know that JS bloat is such a big problem on the web, we should avoid recommending the use of JS for this.

Author: @rkochman What do you think?

Reviewer: Can we recommend the CSS solution, but also note that javascript libraries are available if CSS doesn't meet all their needs?

Author: I'm going to do that and add a note about being careful of bloat.

Author: New language is here: https://github.com/GoogleChrome/web.dev/pull/3215/files#diff-441bd3256a136718e95146958fd6e11eR210

For media content in iframes (such as YouTube videos), try a responsive approach
(like the one [proposed by John
Surdakowski](http://avexdesigns.com/responsive-youtube-embed/)).

{% Aside 'caution' %}
Don't force element sizing that results in an [aspect
ratio](https://www.google.com/search?q=aspect+ratio&oq=aspect+ratio&aqs=chrome..69i57j35i39j0l6.1896j0j7&sourceid=chrome&ie=UTF-8)

> different from the original video. Squashed or stretched looks bad.

Reviewer 2: ```suggestion
different from the original video. Squashed or stretched videos look bad.
```

{% endAside %}

**CSS:**

```css
.video-container {
    position: relative;
    padding-bottom: 56.25%;
    padding-top: 0;
    height: 0;
    overflow: hidden;
}

.video-container iframe,
.video-container object,
.video-container embed {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
}
```

**HTML:**

```html
<div class="video-container">
  <iframe src="//www.youtube.com/embed/l-BA9Ee2XuM"
          frameborder="0" width="560" height="315">
  </iframe>
</div>
```

[Try it](https://googlesamples.github.io/web-fundamentals/fundamentals/media/responsive_embed.html)

Compare the [responsive sample](https://googlesamples.github.io/web-fundamentals/fundamentals/media/responsive_embed.html)
to the [unresponsive version](https://googlesamples.github.io/web-fundamentals/fundamentals/design-and-ux/responsive/unyt.html).

> 

Reviewer 2: Can you add a section mentioning picture-in-picture here? It can just be one paragraph linking to one of Francois's PiP posts.

Author: That requires JavaScript. We decided that all things JS were out of scope for phase I of this section.

### Device orientation

Device orientation isn't an issue for desktop monitors or laptops, but it's
hugely important when considering web page design for mobile devices and
tablets.

Safari on iPhone does a good job of switching between portrait and landscape
orientation:

<div class="w-columns">
<figure class="w-figure">

>   <img src="./iphone-video-playing-portrait.png" alt="Screenshot of video playing in Safari on iPhone, portrait.">

Reviewer 2: ```suggestion
  <img class="w-screenshot w-screenshot--filled" src="./iphone-video-playing-portrait.png" alt="Screenshot of video playing in Safari on iPhone, portrait.">
```

  <figcaption class="w-figcaption">Screenshot of video playing in Safari on iPhone, portrait.</figcaption>
</figure>
<figure class="w-figure">

>   <img src="./iphone-video-playing-landscape.png" alt="Screenshot of video playing in Safari on iPhone, landscape.">

Reviewer 2: ```suggestion
  <img class="w-screenshot w-screenshot--filled" src="./iphone-video-playing-landscape.png" alt="Screenshot of video playing in Safari on iPhone, landscape.">
```

  <figcaption class="w-figcaption">Screenshot of video playing in Safari on iPhone, landscape.</figcaption>
</figure>
</div>

Device orientation on an iPad and Chrome on Android can be problematic.
For example, without any customization a video playing on an iPad in landscape

> orientation looks like this:

Reviewer 2: I'm skeptical about whether this is still true…

<figure class="w-figure">

>   <img src="./ipad-landscape-video-playing.png" alt="Screenshot of video playing in Safari on iPad, landscape.">

Reviewer 2: ```suggestion
  <img class="w-screenshot w-screenshot--filled" src="./ipad-landscape-video-playing.png" alt="Screenshot of video playing in Safari on iPad, landscape.">
```

  <figcaption class="w-figcaption">Screenshot of video playing in Safari on iPad, landscape.</figcaption>
</figure>

Setting the video `width: 100%` or `max-width: 100%` with CSS can resolve
many device orientation layout problems.

### Autoplay

The `autoplay` attribute controls whether the browser downloads and plays a
video immediately. The precise way it works depends on the platform and browser.

* Chrome: Depends on multiple factors including but not limited to whether the
  viewing is on desktop and whether the a mobile user has added your site or app
  to their homescreen. For  details, see [Auotplay Policy
  Changes](https://developers.google.com/web/updates/2017/09/autoplay-policy-changes).

* Firefox: Blocks all video and sound, but gives user the ability to relax these
  restrictions for either all sites or particular sites. For details, see [Allow
  or block media autoplay in Firefox]()

> * Sarfari: Has historically required a user gesterue, but has been relaxing that

Reviewer 2: ```suggestion
* Sarfari: Has historically required a user gesture, but has been relaxing that
```
